google/cloud/datastore/transaction.py

```diff
@@ -240,6 +250,57 @@ def begin(self, retry=None, timeout=None):
             self._status = self._ABORTED
             raise
 
+    def _begin_with_id(self, transaction_id):
+        """
+        Attach newly created transaction to an existing transaction ID.
+
+        This is used when begin_later is True, when the first lookup request
+        associated with this transaction creates a new transaction ID.
+
+        :type transaction_id: bytes
+        :param transaction_id: ID of the transaction to attach to.
+        """
+        if self._status is not self._INITIAL:
+            raise ValueError("Transaction already begun.")
+        self._id = transaction_id
+        self._status = self._IN_PROGRESS
+
+    def _abort_if_not_began(fn: Callable) -> Callable:  # type: ignore
+        """
+        Function wrapper to abort transaction if it hasn't started when
+        the wrapped function is called.
+
+        Used by commit and rollback.
+        """
+
+        @functools.wraps(fn)
+        def wrapped(self, *args, **kwargs):
+            if self._status == self._INITIAL:
+                self._status = self._ABORTED
+                self._id = None
+                return None
+            else:
+                return fn(self, *args, **kwargs)
+
+        return wrapped
+
+    def _begin_if_not_began(fn: Callable) -> Callable:  # type: ignore
+        """
+        Function wrapper to begin transaction if it hasn't started when
+        the wrapped function is called.
+
+        Used by put and delete.
+        """
+
+        @functools.wraps(fn)
+        def wrapped(self, *args, **kwargs):
+            if self._begin_later and self._status == self._INITIAL:
+                self.begin()
+            return fn(self, *args, **kwargs)
+
+        return wrapped
```

Comment:
Maybe we should make the public `id` getter a `_begin_if_not_began` function? Otherwise, existing code will start returning None where it used to have an ID. Thoughts?

Reply:
N/A, begin_later defaulting to False